src/uct/tcp/tcp_ep.c

```diff
@@ -850,6 +971,89 @@ uct_tcp_ep_am_sendv(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
     return status;
 }
 
+static inline void uct_tcp_ep_send_put_ack(uct_tcp_iface_t *iface,
+                                           uct_tcp_ep_t *ep,
+                                           uct_tcp_am_hdr_t *hdr)
+{
+    ucs_assertv(hdr != NULL, "ep=%p", ep);
+
+    hdr->am_id  = UCT_TCP_EP_PUT_ACK_AM_ID;
+    hdr->length = 0;
+    uct_tcp_ep_am_send(iface, ep, hdr);
+}
+
+static ucs_status_t
+uct_tcp_ep_put_ack_handle_pending(uct_pending_req_t *self)
+{
+    uct_tcp_am_hdr_t *hdr = NULL;
+    uct_tcp_ep_put_ack_pending_req_t *put_ack_req;
+    uct_tcp_iface_t *iface;
+    ucs_status_t status;
+
+    put_ack_req = ucs_derived_of(self, uct_tcp_ep_put_ack_pending_req_t);
+    iface       = ucs_derived_of(put_ack_req->ep->super.super.iface,
+                                 uct_tcp_iface_t);
+
+    status = uct_tcp_ep_am_prepare(iface, put_ack_req->ep, 0, &hdr);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    uct_tcp_ep_send_put_ack(iface, put_ack_req->ep, hdr);
+    ucs_free(put_ack_req);
+
+    return UCS_OK;
+}
+
+static void uct_tcp_ep_post_put_ack(uct_tcp_ep_t *ep,
+                                    uct_tcp_ep_put_req_t *put_req)
+{
+    uct_tcp_am_hdr_t *hdr  = NULL;
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    uct_tcp_ep_put_ack_pending_req_t *put_ack_req;
+    ucs_status_t status;
+
+    /* Make sure that we are sending nothing through this EP at the moment.
+     * This check is needed to avoid mixing AM/PUT data sent from this EP
+     * and this PUT ACK message */
+    status = uct_tcp_ep_am_prepare(iface, ep, 0, &hdr);
+    if (ucs_likely(status == UCS_OK)) {
+        uct_tcp_ep_send_put_ack(iface, ep, hdr);
+        return;
+    } else if (status != UCS_ERR_NO_RESOURCE) {
+        ucs_error("tcp_ep %p: failed to prepare AM data", ep);
+        return;
+    }
+
+    ucs_assert(status == UCS_ERR_NO_RESOURCE);
+    
+    /* There are no resources to send this PUT ACK message from
+     * this EP. Add request to a pending queue */
+    put_ack_req = ucs_calloc(1, sizeof(*put_ack_req),
```

Comment:
maybe instead of going through pending queue, we can add a counter per ep (rx side) of how many ACKs should be sent? and send then directly from uct_tcp_ep_progress_data_tx

Reply:
it will increase memory consumption for TCP EP structure.
it seems that this is more complicated logic than having a pending queue or I'm missing smth

Reply:
@yosefe will do it, but it will require additional if statement inside `uct_tcp_ep_progress_data_tx` after EP TX buffer is getting empty to check whether we need to send PUT ACK or not.
wdyt?

Reply:
it's ok since we should have ep on pending queue only for slowpath case anyway

Reply:
> it's ok since we should have ep on pending queue only for slowpath case anyway

@yosefe `uct_tcp_ep_progress_data_tx()` will check it per each `uct_iface_progress()` call - not only when handling pending queue

Reply:
can we do it from pending queue context? like in ud transport

Reply:
> can we do it from pending queue context? like in ud transport

@yosefe thanks for the hint - will check how it is done by UD transport